fdbserver/TLogServer.actor.cpp

```diff
@@ -1757,12 +1756,19 @@ Future<Void> tLogPeekMessages(PromiseType replyPromise,
 		wait(logData->version.whenAtLeast(reqBegin));
 		wait(delay(SERVER_KNOBS->TLOG_PEEK_DELAY, g_network->getCurrentTask()));
 	}
-	if (!logData->stopped && reqTag.locality != tagLocalityTxs && reqTag != txsTag) {
+	if (!logData->stopped && reqTag.locality != tagLocalityTxs && reqTag != txsTag &&
+	    logData->version.get() < logData->recoveryTxnVersion) {
+		DisabledTraceEvent("TLogPeekMessages1", self->dbgid)
+		    .detail("LogId", logData->logId)
+		    .detail("Tag", reqTag.toString())
+		    .detail("ReqBegin", reqBegin)
+		    .detail("Version", logData->version.get())
+		    .detail("RecoveredAt", logData->recoveredAt);
 		// Make sure the peek reply has the recovery txn for the current TLog.
 		// Older generation TLog has been stopped and doesn't wait here.
 		// Similarly during recovery, reading transaction state store
 		// doesn't wait here.
-		wait(logData->recoveryTxnReceived.getFuture());
+		wait(logData->version.whenAtLeast(logData->recoveryTxnVersion));
```

Comment:
An explanation of why the tlog needs to wait till it receives the recovery version would be helpful.

Reply:
This is needed to explicitly ensure the contract that TLog never sends an empty version
between `(recoveredAt, recoveryTransactionVersion)`. Otherwise, the storage server
can fail with `rollbackVersion >= data->storageVersion()` assertion. See https://github.com/apple/foundationdb/pull/6914 for more context.

Reply:
Ok, thanks!